std/protocols/bus.asm

```diff
@@ -0,0 +1,114 @@
+use std::protocols::fingerprint::fingerprint_with_id;
+
+// The main implementation, sends the tuple (id, tuple...) to the bus by adding
+// `multiplicity / (beta - fingerprint(id, tuple...))` to `acc`
+// It is the callers responsibility to properly constrain the multiplicity (e.g. constrain
+// it to be boolean) if needed.
+let bus_interaction: expr, expr, expr[], expr, expr[], Fp2<expr>, Fp2<expr> -> Constr[] = |is_first, id, tuple, multiplicity, acc, alpha, beta| {
+    let _ = if !is_extension(acc) {
+        assert(!needs_extension(), || "The Goldilocks field is too small and needs to move to the extension field. Pass two accumulators instead!")
+    } else { };
+
+    // On the extension field, we'll need two field elements to represent the challenge.
+    // If we don't need an extension field, we can simply set the second component to 0,
+    // in which case the operations below effectively only operate on the first component.
+    let fp2_from_array = |arr| if is_extension(acc) { Fp2::Fp2(arr[0], arr[1]) } else { from_base(arr[0]) };
+    let acc_ext = fp2_from_array(acc);
+
+    let next_acc = if is_extension(acc) {
+        next_ext(acc_ext)
+    } else {
+        // The second component is 0, but the next operator is not defined on it...
+        from_base(acc[0]')
+    };
+
+    // Implemented as: folded = (beta - fingerprint(id, tuple...));
+    // `multiplicity / (beta - fingerprint(id, tuple...))` to `acc`
+    let folded = sub_ext(beta, fingerprint_with_id(id, tuple, alpha));
+    let folded_next = next_ext(folded);
+
+    let m_ext = from_base(multiplicity);
+    let m_ext_next = next_ext(m_ext);
+
+    let is_first_next = from_base(is_first');
+
+    
+    // Update rule:
+    // fingerprint_with_id(id, (a1', a2', ...), alpha) * (acc' - acc * (1 - is_first')) - multiplicity' = 0
+    let update_expr = sub_ext(
+        mul_ext(folded_next, sub_ext(next_acc, mul_ext(acc_ext, sub_ext(from_base(1), is_first_next)))), m_ext_next
+    );
+
+    let (acc_1, acc_2) = unpack_ext(acc_ext);
+    let (expr_1, expr_2) = unpack_ext(update_expr);
+    
+    constrain_eq_ext(update_expr, from_base(0))
+};
+
+let compute_next_z_send: expr, expr, expr[], expr, Fp2<expr>, Fp2<expr>, Fp2<expr> -> fe[] = query |is_first, id, tuple, multiplicity, acc, alpha, beta| {
+    // Implemented as: folded = (beta - fingerprint(id, tuple...));
+    // `multiplicity / (beta - fingerprint(id, tuple...))` to `acc`
+    let folded = sub_ext(beta, fingerprint_with_id(id, tuple, alpha));
+    let folded_next = next_ext(folded);
+
+    let m_ext = from_base(multiplicity);
+    let m_ext_next = next_ext(m_ext);
+
+    let is_first_next = from_base(is_first');
+    
+    // acc' = acc * (1 - is_first') + multiplicity / fingerprint_with_id(id, (a1', a2'))
+    let res = add_ext(
+        mul_ext(eval_ext(acc), eval_ext(sub_ext(from_base(1), is_first_next))),
+        mul_ext(eval_ext(m_ext_next), inv_ext(eval_ext(folded_next)))
+    );
```

Comment:
```suggestion
    let is_first_next = eval(is_first');
    let current_acc = if is_first_next == 1 { from_base(0) } else { eval_ext(acc) };
    
    // acc' = current_acc + multiplicity / fingerprint_with_id(id, (a1', a2'))
    let res = add_ext(
        current_acc,
        mul_ext(eval_ext(m_ext_next), inv_ext(eval_ext(folded_next)))
    );
```

This fixes the extension test! Previously, the hint was never executed, because the argument `acc` was never known. What happens now is this:
1. The hint is run on the *last* row. Because `is_first'` evaluates to `1`, `current_acc == from_base(0)` and `acc` (which is unknown) never needs to be evaluated.
2. This sets `z_next` in the last row, which sets `z` in the the first row.
3. When evaluating the hint on the first row, `is_first'` evaluates to `0`, so `acc` (a.k.a `z`) needs to be evaluated. But we've just set it, so it works.

In your previous code, you would have evaluated `acc` in step 1 and then multiply by `0`. That is correct, but witgen is not smart enough to know that the when multiplying by 0 it doesn't need to evaluate the other factor.

Also, this code should be more efficient ;)

Reply:
Thanks for the explanation 🎉